File: app/src/main/kotlin/xyz/retrixe/salezy/api/entities/Customer.kt
package xyz.retrixe.salezy.api.entities

import kotlinx.serialization.Serializable

@Serializable data class Customer(
    val id: Int,
    val phone: String,
    val name: String?,
    val email: String?,
    val address: String?,
    val taxIdNumber: String?,
    val notes: String?,
)

@Serializable data class EphemeralCustomer(
    val phone: String,
    val name: String?,
    val email: String?,
    val address: String?,
    val taxIdNumber: String?,
    val notes: String?,
)

File: server/src/database/entities/customer.ts
import { ajv } from '../../utils.js'

export interface Customer {
  id: number
  phone: string | null
  name: string | null
  email: string | null
  address: string | null
  taxIdNumber: string | null
  notes: string | null
}

export type EphemeralCustomer = Omit<Customer, 'id'>

export const validateCustomer = ajv.compile<Customer>({
  type: 'object',
  properties: {
    id: { type: 'integer' },
    phone: { type: 'string', maxLength: 20, minLength: 1, nullable: true },
    name: { type: 'string', maxLength: 320, minLength: 1, nullable: true },
    email: { type: 'string', maxLength: 320, minLength: 1, nullable: true },
    address: { type: 'string', nullable: true },
    taxIdNumber: { type: 'string', maxLength: 64, minLength: 1, nullable: true },
    notes: { type: 'string', nullable: true },
  },
  required: ['id'],
  additionalProperties: false,
})

const { ...ephemeralCustomerSchema } = validateCustomer.schema as Record<string, any>
delete ephemeralCustomerSchema.required
delete ephemeralCustomerSchema.properties.id
export const validateEphemeralCustomer = ajv.compile<EphemeralCustomer>(ephemeralCustomerSchema)
